libfreerdp/core/gateway/rdg.c

```diff
@@ -161,20 +161,57 @@ static BOOL rdg_send_tunnel_request(rdpRdg* rdg)
 {
 	wStream* s;
 	BOOL status;
-	s = Stream_New(NULL, 16);
+	WCHAR* PAACookie = NULL;
+    UINT16 PAACookieLen = 0;
+	int i;
 
-	if (!s)
-		return FALSE;
+
+	if (rdg->extAuth == HTTP_EXTENDED_AUTH_PAA)
+	{
+		PAACookieLen = ConvertToUnicode(CP_UTF8, 0, rdg->settings->GatewayAccessToken, -1, &PAACookie, 0);
+		if (!PAACookie)
+			return FALSE;
+
+		s = Stream_New(NULL, 16 + 2 + 2*PAACookieLen);
+		if (!s)
+		{
+			free(PAACookie);
+			return FALSE;
+		}
+	}
+	else
+	{
+		s = Stream_New(NULL, 16);
+
+		if (!s)
+			return FALSE;
+	}
 
 	Stream_Write_UINT16(s, PKT_TYPE_TUNNEL_CREATE); /* Type (2 bytes) */
 	Stream_Write_UINT16(s, 0); /* Reserved (2 bytes) */
 	Stream_Write_UINT32(s, 16); /* PacketLength (4 bytes) */
 	Stream_Write_UINT32(s, HTTP_CAPABILITY_TYPE_QUAR_SOH); /* CapabilityFlags (4 bytes) */
-	Stream_Write_UINT16(s, 0); /* FieldsPresent (2 bytes) */
-	Stream_Write_UINT16(s, 0); /* Reserved (2 bytes), must be 0 */
+
+	if (rdg->extAuth == HTTP_EXTENDED_AUTH_PAA)
+	{
+		Stream_Write_UINT16(s, 1); /* FieldsPresent (2 bytes) */
+		Stream_Write_UINT16(s, 0); /* Reserved (2 bytes), must be 0 */
+
+       	Stream_Write_UINT16(s, PAACookieLen * 2); /* PAA cookie string length */
```

Comment:
Use `Stream_Write_UTF16_String` to write a UTF16 string

Reply:
Ok, done